drivers/platform/stm32/stm32_irq.c

```diff
@@ -0,0 +1,267 @@
+
+	descriptor->extra = extra;
+	*desc = descriptor;
+
+	return SUCCESS;
+error:
+	free(extra);
+	free(desc);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by irq_ctrl_init()
+ * @param desc - IRQ descriptor
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_ctrl_remove(struct irq_ctrl_desc *desc)
+{
+	if (desc) {
+		if (desc->extra)
+			free(desc->extra);
+
+		free(desc);
+		return SUCCESS;
+	}
+
+	return -ENOMEM;
+}
+
+/**
+ * @brief Register a callback to handle the irq events
+ * @param desc - IRQ descriptor
+ * @param irq_id - IRQ ID
+ * @param callback_desc -  IRQ callback descriptor
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_register_callback(struct irq_ctrl_desc *desc, uint32_t irq_id,
+				    struct callback_desc *callback_desc)
+{
+	switch (irq_id) {
+	case EXTERNAL_INT_ID1:
+	case EXTERNAL_INT_ID2:
+	case EXTERNAL_INT_ID3:
+	case EXTERNAL_INT_ID4:
+	case EXTERNAL_INT_ID5:
+		stm32_irq_callback[irq_id].GPIO_Pin = ((stm32_irq_desc *)(
+				desc->extra))->gpio_num;
+		stm32_irq_callback[irq_id].desc.callback = callback_desc->callback;
+		stm32_irq_callback[irq_id].desc.config = callback_desc->config;
+		stm32_irq_callback[irq_id].desc.ctx = callback_desc->ctx;
+		break;
+
+	default:
+		return -EINVAL;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Unregister a callback to handle the irq events
+ * @param desc - IRQ descriptor
+ * @param irq_id - IRQ ID
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_unregister(struct irq_ctrl_desc *desc, uint32_t irq_id)
+{
+	if (!desc)
+		return -ENOMEM;
+
+	if (irq_id >= NB_INTERRUPTS)
+		return -EINVAL;
+
+	stm32_irq_callback[irq_id].desc.callback = NULL;
+	return SUCCESS;
+}
+
+/**
+ * @brief Global interrupt enable
+ * @param desc - IRQ descriptor
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_global_enable(struct irq_ctrl_desc *desc)
+{
+	__enable_irq();
+	return SUCCESS;
+}
+
+/**
+ * @brief Global interrupt disable
+ * @param desc - IRQ descriptor
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_global_disable(struct irq_ctrl_desc *desc)
+{
+	__disable_irq();
+	return SUCCESS;
+}
+
+/**
+ * @brief Specific interrupt enable
+ * @param desc - IRQ descriptor
+ * @param irq_id - IRQ ID
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_enable(struct irq_ctrl_desc *desc, uint32_t irq_id)
+{
+	IRQn_Type irq_type;
+
+	irq_type = ((stm32_irq_desc *)(desc->extra))->irq_type;
+	HAL_NVIC_EnableIRQ(irq_type);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Specific interrupt disable
+ * @param desc - IRQ descriptor
+ * @param irq_id - IRQ ID
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t stm32_irq_disable(struct irq_ctrl_desc *desc, uint32_t irq_id)
+{
+	IRQn_Type irq_type;
+
+	irq_type = ((stm32_irq_desc *)(desc->extra))->irq_type;
+	HAL_NVIC_DisableIRQ(irq_type);
```

Comment:
irq_id is ignored, is this the intention ?

Reply:
I will add a check for validating irq_id

Reply:
it's still strange to me that the irq_id is ignored and i did some research - it seems that for EXTI on stm32 it is not possible to disable a specific interrupt but the type (in this case, all EXTI interrupts).

it would be cool to enable/disable specifi EXTI interrupts by unmasking/masking them but i'm not sure if the HAL has API for that, could you please check ? 

if it's not possible, then i'm ok with the current version, we're somewhat forced to do it this way by the CMSIS HAL

Reply:
ok, looks like we canțt go round it, stm32 EXTI is a bit special so no need to change anything ;)